packages/compass-aggregations/src/modules/pipeline-builder/text-editor.ts

```diff
@@ -0,0 +1,209 @@
+  EditorPreviewFetchSuccess = 'compass-aggregations/pipeline-builder/text-editor/TextEditorPreviewFetchSuccess',
+  EditorPreviewFetchError = 'compass-aggregations/pipeline-builder/text-editor/TextEditorPreviewFetchError',
+  EditorValueChange = 'compass-aggregations/pipeline-builder/text-editor/TextEditorValueChange',
+};
+
+type EditorValueChangeAction = {
+  type: EditorActionTypes.EditorValueChange;
+  pipelineText: string;
+  stageOperators: string[];
+  syntaxErrors: PipelineParserError[];
+};
+
+type EditorPreviewFetchAction = {
+  type: EditorActionTypes.EditorPreviewFetch;
+};
+
+type EditorPreviewFetchSuccessAction = {
+  type: EditorActionTypes.EditorPreviewFetchSuccess;
+  previewDocs: Document[];
+};
+
+type EditorPreviewFetchErrorAction = {
+  type: EditorActionTypes.EditorPreviewFetchError;
+  serverError: MongoServerError;
+};
+
+export type TextEditorState = {
+  pipelineText: string;
+  stageOperators: string[];
+  syntaxErrors: PipelineParserError[];
+  serverError: MongoServerError | null;
+  loading: boolean;
+  previewDocs: Document[] | null;
+};
+
+const INITIAL_STATE: TextEditorState = {
+  pipelineText: '',
+  stageOperators: [],
+  syntaxErrors: [],
+  serverError: null,
+  loading: false,
+  previewDocs: null,
+};
+
+const reducer: Reducer<TextEditorState> = (state = INITIAL_STATE, action) => {
+  if (
+    isAction<EditorValueChangeAction>(
+      action,
+      EditorActionTypes.EditorValueChange
+    ) ||
+    isAction<PipelineModeToggledAction>(
+      action,
+      PipelineModeActionTypes.PipelineModeToggled
+    )
+  ) {
+    return {
+      ...state,
+      pipelineText: action.pipelineText,
+      stageOperators: action.stageOperators,
+      syntaxErrors: action.syntaxErrors,
+    };
+  }
+
+  if (
+    isAction<EditorPreviewFetchAction>(
+      action,
+      EditorActionTypes.EditorPreviewFetch
+    )
+  ) {
+    return {
+      ...state,
+      previewDocs: null,
+      serverError: null,
+      loading: true,
+    };
+  }
+
+  if (
+    isAction<EditorPreviewFetchSuccessAction>(
+      action,
+      EditorActionTypes.EditorPreviewFetchSuccess
+    )
+  ) {
+    return {
+      ...state,
+      serverError: null,
+      syntaxErrors: [],
+      loading: false,
+      previewDocs: action.previewDocs,
+    };
+  }
+
+  if (
+    isAction<EditorPreviewFetchErrorAction>(
+      action,
+      EditorActionTypes.EditorPreviewFetchError
+    )
+  ) {
+    return {
+      ...state,
+      serverError: action.serverError,
+      syntaxErrors: [],
+      loading: false,
+      previewDocs: null,
+    };
+  }
+
+  return state;
+};
+
+export const loadPreviewForPipeline = (
+): PipelineBuilderThunkAction<
+  Promise<void>,
+  EditorPreviewFetchAction |
+  EditorPreviewFetchSuccessAction |
+  EditorPreviewFetchErrorAction
+> => {
+  return async (dispatch, getState, { pipelineBuilder }) => {
+    const {
+      autoPreview
+    } = getState();
+
+    const canRunPipeline = pipelineBuilder.syntaxError.length === 0;
```

Comment:
This should also check whether there is an out stage or not, similar to stage editor. Otherwise we are trying to run an out pipeline it seems (and failing with a server error luckily):

<img width="893" alt="image" src="https://user-images.githubusercontent.com/5036933/197491663-a381b073-adf1-493c-ac96-826eb56b9f3f.png">